### DISCUSSION TITLE
Right way to package Avalonia project for Windows using MSIX

### DISCUSSION THREAD

**Discussion by DmitryBorodiy** [AvaloniaUI/Avalonia#16093] [Q&A]
I have avalonia application with target frameworks monikers,because I need to use WinRT(only on Windows).

```
<PropertyGroup>
  <OutputType>WinExe</OutputType>
  <TargetFrameworks>net8.0-windows10.0.19041.0;net8.0</TargetFrameworks>
  <ApplicationManifest>app.manifest</ApplicationManifest>
  <Nullable>disable</Nullable>
  <PlatformTargets>x86</PlatformTargets>
  <Platforms>x64;x86;</Platforms>
</PropertyGroup>
```
Here is a part of project file. I have error when trying package this app into Windows packaging project. It looks like msix don't know where app executable for Windows is placed. How correct configure WAP and csproj?

Here is an issue on msix repo:
https://github.com/microsoft/msix-packaging/issues/635

**Comment by jack775544** [comment]
I am not sure how do build an MSIX package from inside of Visual Studio like you are doing but I have done it from the command line before.

First you are going to need an app manifest like the one found [here](https://www.advancedinstaller.com/hub/msix-packaging/appxmanifest.html).

Then you are going to have to make sure it gets published in your build by adding this in your csproj file
```xml
<ItemGroup>
	<None Update="AppxManifest.xml" CopyToOutputDirectory="PreserveNewest" />
</ItemGroup>
```

Next you have to publish your application with a command like this `dotnet publish --self-contained -p:PublishSingleFile=true -r win-x64 -f net8.0-windows10.0.19041.0 -c Release`

Then move to the directory that the application was published to and run `makeappx pack /d . /p my-application-name.msix`. Note that makeappx needs to be installed via visual studio.

This will give you an unsigned msix file so you will have to sign it with `signtool` if you are not deploying it to the Windows Store.

EDIT: When I have done this in the past, I was not using WinRT, so I don't know if there is any special requirements to make it work.

**Reply by DmitryBorodiy** [reply]
Maybe I can use msix packaging tool. Can I sign this package with Microsoft Store developer account?

**Reply by jack775544** [reply]
I have not done Microsoft Store deploys personally, but I believe you just have to upload the unsigned msix to the store and it will be signed for you.

**Comment by maxkatz6** [comment]
You also can use [Windows Application Packaging Project](https://learn.microsoft.com/en-us/windows/msix/desktop/desktop-to-uwp-packaging-dot-net)
![image](https://github.com/AvaloniaUI/Avalonia/assets/3163374/6b91b858-f5cf-4632-b50a-cfb8fa2f4913)

>  It looks like msix don't know where app executable for Windows is placed

Might be a problem, if it can't decide which target framework to run. Should be easily solvable by creating a dedicated project for windows (which is recommended either way).

**Reply by DmitryBorodiy** [reply]
Maybe some path exists for building only windows target version when building for MSIX?
